query/src/sql/optimizer/property.rs

```diff
@@ -0,0 +1,72 @@
+// Copyright 2021 Datafuse Labs.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::HashSet;
+
+use crate::sql::common::IndexType;
+
+pub type ColumnSet = HashSet<IndexType>;
+
+pub struct NamedColumn {
+    pub index: IndexType,
+    pub name: String,
+}
+
+#[derive(Default, Clone)]
+pub struct RequiredProperty {
+    required_columns: ColumnSet,
+}
+
+impl RequiredProperty {
+    pub fn create(required_columns: ColumnSet) -> Self {
+        RequiredProperty { required_columns }
+    }
+
+    pub fn required_columns(&self) -> &ColumnSet {
+        &self.required_columns
+    }
+
+    pub fn provided_by(
+        &self,
+        relational_prop: &RelationalProperty,
+        _physical_prop: &PhysicalProperty,
+    ) -> bool {
+        self.required_columns()
+            .is_subset(relational_prop.output_columns())
+    }
+}
+
+#[derive(Default, Clone)]
+pub struct RelationalProperty {
+    output_columns: ColumnSet,
+}
+
+impl RelationalProperty {
+    pub fn create(output_columns: ColumnSet) -> Self {
+        RelationalProperty { output_columns }
+    }
+
+    pub fn output_columns(&self) -> &ColumnSet {
+        &self.output_columns
```

Comment:
We don't need ordering for `output_columns` ?

Reply:
We don't need to track order of attributes in a tuple during optimization, since if two different relational expressions can produce same set of attributes then they can be regarded as logically equivalent.

There is a ordered and named column list stored in `OptimizeContext`, with which we can render the final result set of query. Though I haven't utilize with it yet.

https://github.com/datafuselabs/databend/blob/0f17720ddfe8797a87523063c5a5629697b81d44/query/src/sql/optimizer/optimize_context.rs#L37-L55